backend/dropbox/dropbox.go

```diff
@@ -1597,84 +1597,37 @@ func (o *Object) Open(ctx context.Context, options ...fs.OpenOption) (in io.Read
-	chunkSize := int64(o.fs.opt.ChunkSize)
-	chunks := 0
-	if size >= 0 {
-		chunks = int(size/chunkSize) + 1
-	}
-	in := readers.NewCountingReader(in0)
-	buf := make([]byte, int(chunkSize))
-
-	fmtChunk := func(cur int, last bool) {
-		if chunks == 0 && last {
-			fs.Debugf(o, "Streaming chunk %d/%d", cur, cur)
-		} else if chunks == 0 {
-			fs.Debugf(o, "Streaming chunk %d/unknown", cur)
-		} else if chunks != 1 {
-			fs.Debugf(o, "Uploading chunk %d/%d", cur, chunks)
-		}
-	}
-
-	appendArg := files.UploadSessionAppendArg{
-		Close: chunks == 1,
-	}
-
-	// write the first chunk
-	fmtChunk(1, false)
+	// start upload
 	var res *files.UploadSessionStartResult
-	chunk := readers.NewRepeatableLimitReaderBuffer(in, buf, chunkSize)
 	err = o.fs.pacer.Call(func() (bool, error) {
-		// seek to the start in case this is a retry
-		if _, err = chunk.Seek(0, io.SeekStart); err != nil {
-			return false, nil
-		}
-		arg := files.UploadSessionStartArg{
-			Close: appendArg.Close,
-		}
-		res, err = o.fs.srv.UploadSessionStart(&arg, chunk)
+		res, err = o.fs.srv.UploadSessionStart(&files.UploadSessionStartArg{}, nil)
 		return shouldRetry(ctx, err)
 	})
 	if err != nil {
 		return nil, err
 	}
 
+	chunkSize := int64(o.fs.opt.ChunkSize)
+	chunks := int(math.Ceil(float64(size) / float64(chunkSize)))
+
+	// write chunks
+	in := readers.NewCountingReader(in0)
+	buf := make([]byte, int(chunkSize))
 	cursor := files.UploadSessionCursor{
 		SessionId: res.SessionId,
 		Offset:    0,
 	}
-	appendArg.Cursor = &cursor
+	appendArg := files.UploadSessionAppendArg{Cursor: &cursor}
+	for currentChunk := 1; ; currentChunk++ {
+		cursor.Offset = in.BytesRead()
 
-	// write more whole chunks (if any, and if !batching), if
-	// batching write the last chunk also.
-	currentChunk := 2
-	for {
-		if chunks > 0 {
-			// Size known
-			if currentChunk == chunks {
-				// Last chunk
-				if !batching {
-					// if the size is known, only upload full chunks. Remaining bytes are uploaded with
-					// the UploadSessionFinish request.
-					break
-				}
-				appendArg.Close = true
-			} else if currentChunk > chunks {
-				break
-			}
+		if chunks < 0 {
+			fs.Debugf(o, "Streaming chunk %d/unknown", currentChunk)
 		} else {
-			// Size unknown
-			lastReadWasShort := in.BytesRead()-cursor.Offset < uint64(chunkSize)
-			if lastReadWasShort {
-				// if the size is unknown, upload as long as we can read full chunks from the reader.
-				// The UploadSessionFinish request will not contain any payload.
-				// This is also what we want if batching
-				break
-			}
+			fs.Debugf(o, "Uploading chunk %d/%d", currentChunk, chunks)
 		}
-		cursor.Offset = in.BytesRead()
-		fmtChunk(currentChunk, false)
-		chunk = readers.NewRepeatableLimitReaderBuffer(in, buf, chunkSize)
+
+		chunk := readers.NewRepeatableLimitReaderBuffer(in, buf, chunkSize)
 		err = o.fs.pacer.Call(func() (bool, error) {
 			// seek to the start in case this is a retry
 			if _, err = chunk.Seek(0, io.SeekStart); err != nil {
```

Comment:
Grr, GitHub reviews won't let me write a comment on the correct line but a bit below this you'll see

```go
			// after the first chunk is uploaded, we retry everything
			return err != nil, err
```

This is no longer correct since we aren't doing a "first" block.

It should be something like

```go
if currentChunk == 1 {
     		return shouldRetry(ctx, err)
}
// after the first chunk is uploaded, we retry everything
return err != nil, err
```

Reply:
I thought that `shouldRetry` was mostly needed because first chunk was the part of the `UploadSessionStart` call.  Basically, the assumption is that after the session is started everything is retriable.  If that's not correct, I can change it to the suggested code of only retrying the first chunk (though this kind of a special case looks a bit strange.)

Reply:
I see what you mean. Maybe just the comment is misleading now!